src/middle/Stan_math_signatures.ml

```diff
@@ -4,9 +4,242 @@ open Core_kernel
 open Type_conversion
 open Mir
 
+type dimensionality =
+  | DReal
+  | DInts
+  | DReals
+  | DIntAndReals
+  | D1DeepReals
+  | DArrayAndVectors
+  | DVectors
+  | DVector
+  | DMatrix
+  | DDeepVectorized
+  | DAllBase
+
+type fkind = Lpmf | Lpdf | Rng | Cdf | Ccdf | Functional | DimPromoting
```

Comment:
I find the term `Functional` a bit confusing. Could we just call it `RealValued`?
Also could we document the intent of some of these in comments? Same for `dimensionality`?